### ISSUE TITLE
Game crashes using SDL on MacOS with an external display attached

### ISSUE THREAD

**Issue by ben9583** [ppy/osu-framework#4083]
I'm working on an `osu-framework` project using the SDL graphical backend, but I am unable to start it when I have an external display connected via thunderbolt 3, even if the window is supposed to be on the primary display. I am using a 16-inch Macbook Pro with MacOS Catalina 10.15.7 and a Mobile Pixels Duex Pro external display.

When I run the project without an external display, everything works as expected; after a bit of time, the window appears and displays the content to the screen. This also happens when I have the display connected while `useOsuTK: true`. However, using SDL with the display attached, I get an exception in the `host.run()` that I don't really understand:

Exception: `System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values. (Parameter 'index')`
```
   at System.Char.IsDigit(String s, Int32 index)
   at osu.Framework.Platform.PassthroughGraphicsBackend.<>c.<getVersionNumberSubstring>b__25_0(String s)
   at System.Linq.Enumerable.TryGetFirst[TSource](IEnumerable`1 source, Func`2 predicate, Boolean& found)
   at System.Linq.Enumerable.FirstOrDefault[TSource](IEnumerable`1 source, Func`2 predicate)
   at osu.Framework.Platform.PassthroughGraphicsBackend.getVersionNumberSubstring(String version)
   at osu.Framework.Platform.PassthroughGraphicsBackend.Initialise(IWindow window)
   at osu.Framework.Platform.SDL2.SDL2GraphicsBackend.Initialise(IWindow window)
   at osu.Framework.Platform.SDL2DesktopWindow.Create()
   at osu.Framework.Platform.MacOS.MacOSWindow.Create()
   at osu.Framework.Platform.GameHost.Run(Game game)
   at Piously.Desktop.Program.Main(String[] args) in /Users/ben/Desktop/Git_Stuff/piously_Git/Piously.Desktop/Program.cs:line 14
```

The `program.cs` file where the crash occurs:
```cs
public static class Program
{
    [STAThread]
    public static int Main(string[] args)
    {
        using (GameHost host = Host.GetSuitableHost(@"Piously", useOsuTK: false)) // No issues when useOsuTK is true or no external display
            host.Run(new PiouslyGameDesktop()); // Exception here

        return 0;
    }
}
```

**Comment by bdach** [comment]
I'm not sure what the external display has to do with it, but clearly going off of this code:

https://github.com/ppy/osu-framework/blob/388be39457fd5bf90a02fb65af4825dfb45cbeee/osu.Framework/Platform/PassthroughGraphicsBackend.cs#L124-L130

you must be getting an [empty OpenGL version string](https://github.com/ppy/osu-framework/blob/388be39457fd5bf90a02fb65af4825dfb45cbeee/osu.Framework/Platform/PassthroughGraphicsBackend.cs#L46), which I'm not even necessarily sure how to debug why would that be the case (possibly calling `GL_GetError()`?). Certainly sounds like there's a probability this isn't the framework's problem.

**Comment by ben9583** [comment]
After some testing, I learned that while running SampleGame in osu-framework does not appear to crash, running osu.Framework.Tests results in the crash, and both are with SDL. I am still not sure why this is the case.

**Comment by peppy** [comment]
The only thing I can think of here is that the one which is failing has a different display stored to the config, which is somehow inhibiting the ability to get a gl context. Could you check the `framework.ini` files for each of the projects you're running and see if that correlates to the error?

**Comment by ben9583** [comment]
Ok, I have more info regarding specifically when the crash happens.

This is the layout of my displays in which the program crashes on initializing:
<img width="171" alt="Screen Shot 2020-12-10 at 11 12 24 AM" src="https://user-images.githubusercontent.com/16968917/101812450-9873b980-3ad8-11eb-8e35-72ffc3faab5c.png">

The primary display is the horizontal screen on the left, while the external display is the vertical one on the right. This screen layout causes the game to crash when the window is at the maximum allowed space (display size minus dock at the bottom) from the previous time played. However, if I change my layout to that of the next images, there is no crash when the window was previously at the maximum window size and I try to relaunch it.

Does not crash with these layouts:
<img width="121" alt="Screen Shot 2020-12-10 at 11 14 43 AM" src="https://user-images.githubusercontent.com/16968917/101812690-ebe60780-3ad8-11eb-9e22-65607bf8f2b1.png">
<img width="155" alt="Screen Shot 2020-12-10 at 11 17 04 AM" src="https://user-images.githubusercontent.com/16968917/101812916-467f6380-3ad9-11eb-9c3f-a4d5de9ec661.png">
<img width="116" alt="Screen Shot 2020-12-10 at 11 17 11 AM" src="https://user-images.githubusercontent.com/16968917/101812927-48e1bd80-3ad9-11eb-8d3b-f5c6ab255ec1.png">

I do not know what this has to do with OpenGL, but these are my findings. For now, I'll be having my external display be attached to the left side of my primary display since this doesn't cause the crash.

**Comment by bdach** [comment]
An interesting detail for sure, but I'm not sure I can glean much from the fact that it only seems to not work on this one side... As a next step I would probably poke the size/position code in `SDL2DesktopWindow` and see if there aren't any obviously wrong values.

**Comment by peppy** [comment]
@ben9583 this is my screen layout, and i cannot reproduce a crash:

![image](https://user-images.githubusercontent.com/191335/102329192-717f2100-3fcb-11eb-89de-fce2195681b2.png)

Are you able to try going back to your old layout and repositioning to be more centered? Also can you confirm whether your game window is starting on your main or secondary screen in the fail case?

**Comment by ben9583** [comment]
So weird thing happened: When I updated my MacBook, my kernel extensions got messed up, so I had to reinstall the drivers for my display. The display was in landscape mode and the option to rotate it was removed. I haven't been able to figure out how to make it portrait mode like it was previously and so it has just been in landscape mode. No crashes occur in this case, even when the display was to the right of my primary.

My guess is that this was an issue with the display, not `osu-framework`. Disappointed I can't work with it in portrait mode, but that's another problem, so I think it's best this issue be closed. By the way, back when the program was crashing, the `framework.ini` showed that the window was starting on my primary display.
